pkg/teeth/sql/publisher_source.sql

```diff
@@ -0,0 +1,42 @@
+SELECT
+  pull_request_events.delivery_id,
+  delivery_events.pull_request_id,
+  html_url AS pull_request_html_url,
+  delivery_events.received,
+  logs_uri,
+  head_sha
+FROM
+  `%s` AS pull_request_events
+JOIN (
+  SELECT
+    delivery_id,
+    received,
+    logs_uri,
+    SAFE.INT64(pull_request.id) AS pull_request_id,
+    LAX_STRING(pull_request.url) AS pull_request_url,
+    LAX_STRING(events.payload.workflow_run.head_sha) AS head_sha,
+  FROM
+    `%s` leech_status
```

Comment:
I'm not a big fan of string interpolation inside a SQL query, though I understand the threat model here isn't as big, as these should only be set up by a trusted admin.

I was able to get around string interpolation by using unqualified identifiers and setting the default project and dataset id in the bigquery client's Query struct.

Reply:
I looked into this as I was wondering why the retry service and leech pipeline also use string interpolation. 

From https://cloud.google.com/bigquery/docs/parameterized-queries#go,
> Parameters cannot be used as substitutes for identifiers, column names, table names, or other parts of the query.

Apparently it just won't expand. An alternative is https://cloud.google.com/bigquery/docs/querying-wildcard-tables#wildcard_table_syntax, but that's not supported for views.

Reply:
You can get around the wildcard issue by just using unqualified table names and setting the project id and dataset id in your client.

https://cloud.google.com/bigquery/docs/reference/standard-sql/data-definition-language#table_path

So you would do something like

```sql
SELECT *
FROM leech_table
```

That way we don't need to inject any values into the query.

Reply:
Unfortunately the beam BigQuery connector doesn't support those features, which was one of the reasons I wanted to not use beam for this pipeline.

Reply:
Another option would be to validate the inputs. 

Project id: 6 to 30 lowercase letters, digits, or hyphens. It must start with a letter. Trailing hyphens are prohibited.
https://cloud.google.com/resource-manager/reference/rest/v1/projects

Dataset ID: up to 1,024 characters, uppercase,lowrcase, numbers and underscores
https://cloud.google.com/bigquery/docs/datasets

Reply:
See the new commit with templating. And yes, as agreed, we are not pursuing BEAM in favor of Cloud Run.

Reply:
That's looking better, though just wondering if we really need interpolation here. I don't really see a user wanting to change the table names, so I was imagining just hard coding them into the query.

Reply:
The table names are dependent on the job's configuration at startup. There has to be a way to configure them.

e.g,
Retry Service (Cloud Run): https://github.com/abcxyz/github-metrics-aggregator/blob/main/pkg/webhook/config.go#L134

Leech (BEAM): https://github.com/abcxyz/github-metrics-aggregator/blob/main/pipeline/leech/main.go#L89-L93

Reply:
I see, that seems like more complexity than we need, but doesn't make sense to revisit those. I would just validate them then.

Reply:
A lot of this should already be done in the CI/CD pipeline.

Autopush: https://github.com/abcxyz/github-metrics-aggregator/blob/5032fb9519922e0dbea530f44d50ae9df65a8103/.github/workflows/ci.yml#L309
Prod: https://github.com/abcxyz/infra-gcp/blob/d741f2ca25450c8e71fa19a053f0ef2b62e89c97/.github/workflows/deploy-github-metrics-aggregator-prod.yml#L180

Now why they are in different repos ... 😕 no idea.

Reply:
I'll probably do a separate PR to validate in this and the existing repos, add a pkg or something.